### PR TITLE
Addition of GroupNormalization

### PR DESCRIPTION
**Pull request by titu1994**
Adds Group Normalization + associated tests from the paper [Group Normalization](https://arxiv.org/abs/1803.08494)

By extention, setting the parameter "groups" to 1 should make it equivalent to Instance Normalization and settng "groups" to the number of input channels will apply layer normalization.

### CODE REVIEW

**Comment by titu1994** [comment]
CNTK tests fail due to library issues, however the Keras code for mean and standard calculation remains the same for all backends, so I feel it can be skipped in the short term.

**Comment by ahundt** [comment]
This looks nice. I bet it is something that is comparatively likely to be accepted into Keras upstream. Should we check for interest there first?

**Comment by titu1994** [comment]
@ahundt Sure, I'll add a request there. For the time being however, I think I'll have to switch to manually computing the mean and variance if I push to Keras. It doesnt have the equivalent K.moments() op, which we added to Contrib quite some time ago. 

There are performance repercussions in this implementation. While it follows the code from the paper quite closely (with some modifications for how Keras handles different data formats etc), K.batch_normalization() and the internal code of batchnorm checks for and uses fused batchnorm ops when available. 

Those won't be available for either contrib or Keras proper until the appropriate ops are added to TF itself (and all this is under the hope that GroupNormalization gains enough favor in the community to warrant an additional fused op in TF). 

Edit: I forgot to say, I think we can keep this version here, since it uses the contrib `moments` op anyway.

File: tests/keras_contrib/backend/backend_test.py [hunk 1]
```diff
@@ -151,14 +153,21 @@ def test_moments(self):
...
                     tf_mean_val = KTF.eval(tf_mean)
+                    # cntk_mean_val = KCTK.eval(cntk_mean)
                     th_var_val = KTH.eval(th_var)
                     tf_var_val = KTF.eval(tf_var)
+                    # cntk_var_val = KCTK.eval(cntk_var)
 
                     # absolute tolerance needed when working with zeros
                     assert_allclose(th_mean_val, tf_mean_val, rtol=1e-4, atol=1e-10)
                     assert_allclose(th_var_val, tf_var_val, rtol=1e-4, atol=1e-10)
+                    # assert_allclose(th_mean_val, cntk_mean_val, rtol=1e-4, atol=1e-10)
```
Comment (ahundt) [inline tests/keras_contrib/backend/backend_test.py]: should probably correct the commented lines in this file

**Comment by titu1994** [comment]
Can't builds won't pass, either here or in Keras proper. I'm investigating the reason currently but have no plausible source of error yet.

**Comment by 0x00b1** [comment]
Thanks, @titu1994. I started merging this into 0.2.0 of [Keras-ResNet](https://github.com/broadinstitute/keras-resnet):

https://github.com/broadinstitute/keras-resnet/pull/40

Would you mind if I added you to the AUTHORS.rst?

**Comment by titu1994** [comment]
Sure, I don't mind, but know that grouped convolutions as I have implemented here do not work on CNTK backend. TF and Theano show no problems, and I've been using this frequently for the past few weeks.

I would say, unless you are fine tuning a task, or have small batch sizes, this is inferior to batch norm + 128 batch size. That's just my observation on very select tests though so I won't comment on overall benefit.

**Comment by 0x00b1** [comment]
@titu1994 I’ll work on fixing the CNTK issues. I’ve actually seen worse results with typical batch sizes, so it won’t replace `BatchNormalization` as the default normalization layer. However, the Keras-ResNet is used by a few packages for image segmentation and object detection where batch sizes are usually one or two and group normalization performs **dramatic** performance benefits.

**Comment by titu1994** [comment]
Yeah I've noticed similar results. It's surprisingly powerful for small batch sizes, even beating instance norm with just groups of 2. It's weird that it would have such a significant difference. 

I'm any case, I've been using it as a drop in replacement for instance norm in a few style transfer tasks, and it performs on par with actual instance norm at 16 batch size so it's useful enough for me.

**Comment by titu1994** [comment]
Btw, if you do get a fix for the CNTK issues, please send a fix here as well.

**Comment by 0x00b1** [comment]
@titu1994 Of course. 👍 Ideally, it won’t stay in Keras-ResNet for too long.

File: keras_contrib/layers/normalization.py [hunk 2]
```diff
@@ -373,4 +374,186 @@ def get_config(self):
...
+
+class GroupNormalization(Layer):
+    """Group normalization layer
+
+    Group Normalization divides the channels into groups and computes within each group
+    the mean and variance for normalization. GN's computation is independent of batch sizes,
+    and its accuracy is stable in a wide range of batch sizes
+
+    # Arguments
+        groups: Integer, the number of groups for Group Normalization.
```
Comment (ahundt) [inline keras_contrib/layers/normalization.py]: Can this include advice on how to choose the integer?

File: tests/keras_contrib/backend/backend_test.py (hunk 1 above)
Comment (titu1994) [reply tests/keras_contrib/backend/backend_test.py]: The issue is that CNTK tests won't pass for some reason. It is using Contrib's `K.moments()` call, and technically, it should do just fine on CNTK, but it fails. And now that Contribs tests are broken, we cant check properly either.

File: keras_contrib/layers/normalization.py (hunk 2 above)
Comment (titu1994) [reply keras_contrib/layers/normalization.py]: Hmm, I can mention that setting groups to 1 makes this identical to Layer Normalization, and setting it to the number of input channels makes this Instance Normalization?

Comment (titu1994) [reply keras_contrib/layers/normalization.py]: Potential fix, use the private `_moments` method directly from the CNTK backend. It will be more fragile to internal Keras changes, but I think moments op is too general to have any major changes.

**Comment by titu1994** [comment]
@ahundt I may have a potential solution to the CNTK issue, which is to use the private backend function `_moments` from the CNTK backend directly and wrap it in `KC.moments()`.

Also, added the comments about how altering `groups` will affect the behavior of the normalization.

**Comment by titu1994** [comment]
@ahundt I dunno if its just my setup or it's an actual issue, but Keras 2.2.0 combined with TF 1.8 always segfaults when I use K.clear_session() and follow up with building another model and then calling fit.

Perhaps that is another source of test errors. I feel we should create a PR that mimics Keras and get these tests passing at least.

**Comment by titu1994** [comment]
@ahundt This is ready for merging as well, and passes CNTK tests now.

Are there any other corrections I should make, or can we merge this?
